### PR TITLE
Abstract over internal deframer state

### PR DESCRIPTION
**Pull request by djc**
Clarifies the deframer interface somewhat.

### CODE REVIEW

**Comment by briansmith** [comment]
I think we should work together on a requirements/design doc for the ultimate state we want to get to w.r.t. deframer and buffer handling, and try to make an efficient plan about how to get there. There are a lot of things that need to be improved w.r.t. these facets and it would be good to ensure we have agreement on what to do before making changes.

**Comment by djc** [comment]
Sure, that makes sense. I no longer have as much time to spend on this, so I'm not sure I'll be able to do much writing on that document.

**Review by ctz — APPROVED** [review]

File: rustls/src/msgs/deframer.rs [hunk 1]
```diff
@@ -103,6 +87,25 @@ impl MessageDeframer {
             self.used = 0;
         }
 
+        Ok(self.frames.pop_front())
+    }
+
+    /// Read some bytes from `rd`, and add them to our internal
+    /// buffer.  If this means our internal buffer contains
+    /// full messages, decode them all.
```
[the remark this answers is not in the document]
Comment (djc) [reply rustls/src/msgs/deframer.rs]: Ah, good call. Fixed that up and added some comments for `pop()`.

File: rustls/src/msgs/deframer.rs [hunk 2]
```diff
@@ -46,25 +46,12 @@ impl MessageDeframer {
     /// Returns an `Error` if the deframer failed to parse some message contents,
     /// `Ok(None)` if no full message is buffered, and `Ok(Some(_))` if a valid message was found.
     pub fn pop(&mut self) -> Result<Option<OpaqueMessage>, Error> {
-        match self.desynced {
-            false => Ok(self.frames.pop_front()),
-            true => Err(Error::CorruptMessage),
+        if self.desynced {
+            return Err(Error::CorruptMessage);
+        } else if let Some(msg) = self.frames.pop_front() {
```
Comment (djc) [inline rustls/src/msgs/deframer.rs:51]: Added a fast path here where we don't try to parse again if there are still parsed messages in the queue.

**Pull request merged**